examples/src/main/scala/gpt/v2.scala

```diff
@@ -0,0 +1,1012 @@
+// cSpell: ignore nn, probs, numel, itemsize, nbytes
+// cSpell: ignore xbow, xprev, isinstance, idx, tok_emb
+// cSpell: ignore Andrej, Karpathy
+
+import java.nio.file.Paths
+import java.nio.file.Files
+import java.net.URL
+import java.net.URI
+
+import scala.annotation.targetName
+import scala.util.Random
+import scala.util.Using
+import scala.collection.immutable.SortedSet
+
+import org.bytedeco.pytorch.OutputArchive
+import org.bytedeco.javacpp.PointerScope
+
+import torch.*
+import torch.Device.CUDA
+import torch.Device.CPU
+import torch.nn.functional as F
+import torch.nn.modules.Module
+import torch.nn.modules.HasParams
+import torch.{---, Slice}
+import torch.optim.Adam
+import torch.DType.float32
+import org.bytedeco.javacpp.Pointer
+import torch.nn.modules.HasWeight
+import org.bytedeco.pytorch.cuda.Stat
+import org.bytedeco.pytorch.cuda.CheckpointDelta
+import org.bytedeco.pytorch.cuda.SnapshotInfo
+import org.bytedeco.pytorch.cuda.CUDAAllocator
+import org.bytedeco.pytorch.cuda.SegmentInfo
+import org.bytedeco.pytorch.cuda.BlockInfo
+import org.bytedeco.pytorch.cuda.DeviceStats
+import org.bytedeco.javacpp.BoolPointer
+import org.bytedeco.pytorch.global.torch_cuda
+
+
+
+/**
+  * This is code translated from Andrej Karpathy's [[video https://www.youtube.com/watch?v=kCc8FmEb1nY]]
+  * titled "Let's build GPT: from scratch, in code, spelled out." The video also provides
+  * links to:
+  * * [[A Google colab for the video https://colab.research.google.com/drive/1JMLa53HDuA-i7ZBmqV7ZnA3c_fvtXnx-?usp=sharing]]
+  * * [[GitHub repo for the video https://github.com/karpathy/ng-video-lecture]]
+  * This code tries to replicate the [[original PyTorch code https://github.com/karpathy/ng-video-lecture/blob/master/gpt.py]]
+  * 
+  * Unfortunately the results are not on par with the original. The code works best when 
+  * weight initialization is **not** used. The number of epochs learned must be significantly 
+  * larger (2 orders of magnitude) above the original. 
+  * As per the `gpt.py` code, weight initialization produces much worse results. 
+  * 
+  * @see https://github.com/karpathy/ng-video-lecture/blob/master/gpt.py
+  */
+object V2:
+
+  // Utility functions
+
+  def moduleName(m: Module): String =
+      //m.getClass().getSimpleName()
+      m.toString()
+
+  def moduleClass(m: Module): String =
+      m.getClass().getSimpleName()
+
+  /**
+    * Collects information of a module and returns this as a string. Complex modules
+    * are shown hierarchically. Information includes the modules `toString` output
+    * that usually holds the variable name and class parameter values. We also add
+    * the number of tensor parameter amd their value in the leaf modules. For the 
+    * other modules the sum of the number of tensor parameters are shown. 
+    * 
+    * Use this output to "debug" your networks
+    *
+    * @param m
+    * @return string
+    */
+  def doModuleInfoString(m:Module, indent: Int): String =
```

Comment:
This looks pretty useful in general. Do you think we could move that into `Module`?